### ISSUE TITLE
 saving and loading yolov5 model

### ISSUE THREAD

**Issue by care55** [ultralytics/yolov5#2922]
## ❔Question
hi ,
I'm trying to save my trained model in yolov5 to load it in another session and trained the model from the epoch it stopped how can I save this in a  model  

!python train.py --img 416 --batch 16 --epochs 1 --data '../data.yaml' --cfg ./models/custom_yolov5s.yaml --weights '' --name yolov5s_results  --cache

to call it in:
torch.save(model.state_dict(), path)
and after saving it how can I load it?

## Additional context
note : I want to save my work in drive and load it from it too
I used this code to train my data:
https://colab.research.google.com/drive/1gDZ2xcTOgR39tGGs-EZ6i3RTs16wmzZQ#scrollTo=wbvMlHd_QwMG
thanks...

**Comment by care55** [comment]
@glenn-jocher  can you help me please?

**Comment by wudashuo** [comment]
models are automatically saved in `runs/expXX/weights/`, if you want to use it in another place, you can load it by `model = torch.hub.load('ultralytics/yolov5', 'custom', path_or_model='your.pt')`

**Comment by pravastacaraka** [comment]
@care55 maybe you can try `!python train.py --resume`

**Comment by care55** [comment]
Thanks @wudashuo 
I try it but I have this error
![Screenshot (3935)](https://user-images.githubusercontent.com/83128663/115990933-aaa94c80-a5ce-11eb-9457-53074435938b.png)
![Screenshot (3936)](https://user-images.githubusercontent.com/83128663/115990937-ac731000-a5ce-11eb-9431-7cd17d4a4fb0.png)

**Comment by care55** [comment]
@pravastacaraka 
!python train.py --resume runs/train/yolov5s_results/weights/last.pt
 I tried this but when I trained a lot of epochs  maybe 2000 or more  all cells after a period of time  shutted down for no reason  so I have to restart all cells to get the runs file and all of my work is lost

How can I fix this issue?
can I fix this problem by using  !python train.py --resume without specify the path?

**Comment by glenn-jocher** [comment]
@care55 if your training was interrupted for any reason you may continue where you left off using the `--resume` command. If your training fully completed then you can start a new training starting from a fully trained model using the `--weights` command. Examples:

## Resume Single-GPU

You may not change settings when resuming, and **no additional arguments** other than `--resume` should be passed:
```bash
python train.py --resume  # automatically find latest checkpoint (searches yolov5/ directory)
python train.py --resume path/to/last.pt  # specify resume checkpoint
```

## Resume Multi-GPU

[Multi-GPU DDP trainings](https://docs.ultralytics.com/yolov5/tutorials/multi_gpu_training) must be resumed with the same GPUs and DDP command, i.e. assuming 8 GPUs:
```bash
python -m torch.distributed.launch --nproc_per_node 8 train.py --resume  # resume latest checkpoint
python -m torch.distributed.launch --nproc_per_node 8 train.py --resume path/to/last.pt  # specify resume checkpoint
```

## Start from Pretrained

If you would like to **start training** from a fully trained model, use the `--weights` argument, not the `--resume` argument:
```bash
python train.py --weights path/to/best.pt  # start from pretrained model
```

Good luck and let us know if you have any other questions!

**Comment by care55** [comment]
@glenn-jocher 
Thanks ,so now  I have to put instead of quotation in --weights '' the path of my trained weights like this:
!python train.py --img 416 --batch 16 --epochs 1 --data '../data.yaml' --cfg ./models/custom_yolov5s.yaml --weights path/to/best.pt --name yolov5s_results --cache
 Right?

**Comment by glenn-jocher** [comment]
@care55 yes, though note that if you specify --weights then --cfg is not needed.

**Comment by care55** [comment]
I appreciate your help, that works Thanks @glenn-jocher

**Comment by care55** [comment]
@glenn-jocher 
I have more questions if you can help please ...
Can I train yolov5 model with coco weights ?and if I could how to make this?
Can you please suggest  the best weights to train yolov5 model for object detection?

**Comment by glenn-jocher** [comment]
@care55 yes you can start training from any pretreind YOLOv5 weights using the --weights command. I would recommend you start from the Train Custom Data tutorial which answers many of these questions:

## YOLOv5 Tutorials

* [Train Custom Data](https://docs.ultralytics.com/yolov5/tutorials/train_custom_data)&nbsp; 🚀 RECOMMENDED
* [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/)&nbsp; ☘️ RECOMMENDED
* [Weights & Biases Logging](https://github.com/ultralytics/yolov5/issues/1289)&nbsp; 🌟 NEW
* [Supervisely Ecosystem](https://docs.ultralytics.com/yolov5/tutorials/model_export8)&nbsp; 🌟 NEW
* [Multi-GPU Training](https://docs.ultralytics.com/yolov5/tutorials/multi_gpu_training)
* [PyTorch Hub](https://docs.ultralytics.com/yolov5/tutorials/pytorch_hub_model_loading)&nbsp; ⭐ NEW
* [ONNX and TorchScript Export](https://docs.ultralytics.com/yolov5/tutorials/model_export)
* [Test-Time Augmentation (TTA)](https://docs.ultralytics.com/yolov5/tutorials/test_time_augmentation)
* [Model Ensembling](https://docs.ultralytics.com/yolov5/tutorials/model_ensembling)
* [Model Pruning/Sparsity](https://docs.ultralytics.com/yolov5/tutorials/model_pruning_and_sparsity)
* [Hyperparameter Evolution](https://docs.ultralytics.com/yolov5/tutorials/hyperparameter_evolution)
* [Transfer Learning with Frozen Layers](https://docs.ultralytics.com/yolov5/tutorials/transfer_learning_with_frozen_layers)&nbsp; ⭐ NEW
* [TensorRT Deployment](https://github.com/wang-xinyu/tensorrtx)

**Comment by john8822** [comment]
So @glenn-jocher  can I train yolov5x the first time for 100 epochs and save it in drive then pass these weights in --weight and train it with another 100 epochs  so it becomes 200 epochs because every 50 epochs in yolov5x take almost two hours to complete and I can't keep my laptop opened all the day and the gpu will be full, is that possible or the training must be continue without interrupting?

Because my data is about 4000 images and I tried to do this with yolov5x ...I have reached to 400 epochs by this technic but I don't find much improvement than 50 epochs when I see my images how many epochs I have to train to reach a good accuracy ?
And is there any technic while training to see that my model is improving or only from the images I can see the results ?
Thanks....

**Comment by john8822** [comment]
At the first time I trained with yolov5x weights like this :
!python train.py --img 416 --batch 16 --epochs 100 --data '../data.yaml' --cfg ./models/yolov5x.yaml --weights yolov5x.pt --name yolov5x_results 

At the second time I tried this:
!python train.py --img 416 --batch 16 --epochs 100 --data '../data.yaml'  --weights path/to/best.pt --name yolov5x_results 

Are these instructions right?

**Comment by glenn-jocher** [comment]
@john8822 if your training was **interrupted** for any reason you may continue where you left off using the `--resume` argument. If your training fully completed, you can start a **new** training from any model using the `--weights` argument. Examples:

## Resume Single-GPU

You may not change settings when resuming, and **no additional arguments** other than `--resume` should be passed, with an optional path to the checkpoint you'd like to resume from. If no checkpoint is passed the most recently updated `last.pt` in your `yolov5/` directory is automatically found and used:
```bash
python train.py --resume  # automatically find latest checkpoint (searches yolov5/ directory)
python train.py --resume path/to/last.pt  # specify resume checkpoint
```

## Resume Multi-GPU

[Multi-GPU DDP trainings](https://docs.ultralytics.com/yolov5/tutorials/multi_gpu_training) must be resumed with the same GPUs and DDP command, i.e. assuming 8 GPUs:
```bash
python -m torch.distributed.launch --nproc_per_node 8 train.py --resume  # resume latest checkpoint
python -m torch.distributed.launch --nproc_per_node 8 train.py --resume path/to/last.pt  # specify resume checkpoint
```

## Start from Pretrained

If you would like to **start training** from a fully trained model, use the `--weights` argument, not the `--resume` argument:
```bash
python train.py --weights path/to/best.pt  # start from pretrained model
```

Good luck and let us know if you have any other questions!

**Comment by john8822** [comment]
@glenn-jocher Is there any technic while training to see that my model is improving or only when training is completed I can see the result from the images?

**Comment by glenn-jocher** [comment]
@john8822 see W&B Logging tutorial:

## YOLOv5 Tutorials

* [Train Custom Data](https://docs.ultralytics.com/yolov5/tutorials/train_custom_data)&nbsp; 🚀 RECOMMENDED
* [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/)&nbsp; ☘️ RECOMMENDED
* [Weights & Biases Logging](https://github.com/ultralytics/yolov5/issues/1289)&nbsp; 🌟 NEW
* [Supervisely Ecosystem](https://docs.ultralytics.com/yolov5/tutorials/model_export8)&nbsp; 🌟 NEW
* [Multi-GPU Training](https://docs.ultralytics.com/yolov5/tutorials/multi_gpu_training)
* [PyTorch Hub](https://docs.ultralytics.com/yolov5/tutorials/pytorch_hub_model_loading)&nbsp; ⭐ NEW
* [ONNX and TorchScript Export](https://docs.ultralytics.com/yolov5/tutorials/model_export)
* [Test-Time Augmentation (TTA)](https://docs.ultralytics.com/yolov5/tutorials/test_time_augmentation)
* [Model Ensembling](https://docs.ultralytics.com/yolov5/tutorials/model_ensembling)
* [Model Pruning/Sparsity](https://docs.ultralytics.com/yolov5/tutorials/model_pruning_and_sparsity)
* [Hyperparameter Evolution](https://docs.ultralytics.com/yolov5/tutorials/hyperparameter_evolution)
* [Transfer Learning with Frozen Layers](https://docs.ultralytics.com/yolov5/tutorials/transfer_learning_with_frozen_layers)&nbsp; ⭐ NEW
* [TensorRT Deployment](https://github.com/wang-xinyu/tensorrtx)
